cmd/snap/cmd_run.go

```diff
@@ -530,6 +531,62 @@ func migrateXauthority(info *snap.Info) (string, error) {
 	return targetPath, nil
 }
 
+func activateXdgDocumentPortal(info *snap.Info) error {
+	// Don't do anything for snaps that don't plug the desktop interface
+	//
+	// NOTE: This check is imperfect because we don't really know
+	// if the interface is connected or not but this is an
+	// acceptable compromise for not having to communicate with
+	// snapd in snap run. In a typical desktop session the
+	// document portal can be in use by many applications, not
+	// just by snaps, so this is at most, pre-emptively using some
+	// extra memory.
+	plugsDesktop := false
+	for _, plug := range info.Plugs {
+		if plug.Interface == "desktop" {
+			plugsDesktop = true
+			break
+		}
+	}
+	if !plugsDesktop {
+		return nil
+	}
+
+	// If there is no session bus, our job is done.  We check this
+	// manually to avoid dbus.SessionBus() auto-launching a new
+	// bus.
+	busAddress := osGetenv("DBUS_SESSION_BUS_ADDRESS")
+	if len(busAddress) == 0 {
+		return nil
+	}
+
+	conn, err := dbus.SessionBus()
```

Comment:
As far as checking whether it is running.  One possibility would be to try and process `/proc/self/mountinfo` looking for the expected mount point.  That sounds like it might be worth doing.

As for detecting the absence of xdg-document-portal, that is more difficult.  This would require parsing every `.service` file in every directory holding activation data for the session bus looking for a file matching the document portal bus name.  And we can't even determine every service directory by reading `session.conf`, since the `<standard_session_servicedirs />` references paths compiled into dbus-daemon.

It is possible to query the session bus for its idea of whether the name is activatable, but that won't be any faster than just trying to perform the method call.

Reply:
Does the desktop portal have standard executable name across distributions?  Could we stat the exact known locations for the binary to see if there's any chance the service would run?